src/centralized_telescope.rs

```diff
@@ -0,0 +1,421 @@
+        mut limit: u64,
+    ) -> (u64, Option<Self>) {
+        if round.s_list.len() as u64 == setup.u {
+            let proof_opt = if Self::h2(setup, round) {
+                Some(Self {
+                    v: round.v,
+                    t: round.t,
+                    items: round.s_list.clone(),
+                })
+            } else {
+                None
+            };
+            return (limit, proof_opt);
+        }
+
+        for &s in &bins[round.h_u64 as usize] {
+            if limit == setup.b {
+                return (limit, None);
+            }
+            if let Some(r) = Round::update(round, s, setup.sec_param) {
+                let (l, proof_opt) = Self::dfs(setup, bins, &r, limit.saturating_add(1));
+                if proof_opt.is_some() {
+                    return (l, proof_opt);
+                }
+                limit = l;
+            }
+        }
+        (limit, None)
+    }
+
+    /// Indexed proving algorithm, returns the total number of DFS calls done
+    /// to find a proof and Some(proof) if found within setup.b calls of DFS,
+    /// otherwise None
+    fn prove_index(setup: &Setup, set: &[Element], v: u64) -> (u64, Option<Self>) {
+        let mut bins: Vec<Vec<Element>> = Vec::with_capacity(setup.n_p as usize);
+        for _ in 0..setup.n_p {
+            bins.push(Vec::new());
+        }
+        // Take only up to 2*np elements for efficiency
+        for &s in set.iter().take(setup.n_p.saturating_mul(2) as usize) {
+            match Self::h0(setup, v, s) {
+                Some(h) => {
+                    bins[h as usize].push(s);
+                }
+                None => return (0, None),
+            }
+        }
+
+        let mut limit = 0;
+        for t in 0..setup.d {
+            if limit == setup.b {
+                return (limit, None);
+            }
+            if let Some(r) = Round::new(v, t, setup.n_p, setup.sec_param) {
+                let (l, proof_opt) = Self::dfs(setup, &bins, &r, limit.saturating_add(1));
+                if proof_opt.is_some() {
+                    return (l, proof_opt);
+                }
+                limit = l;
+            }
+        }
+        (limit, None)
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Calls up to setup.r times the prove_index function and returns an empty
+    /// proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &[Element]) -> Option<Self> {
+        (0..setup.r).find_map(|v| Self::prove_index(setup, set, v).1)
+    }
+
+    /// Alba's verification algorithm, follows proving algorithm by running the
+    /// same depth-first search algorithm.
+    pub fn verify(setup: &Setup, proof: &Self) -> bool {
+        if proof.t >= setup.d || proof.v >= setup.r || proof.items.len() as u64 != setup.u {
+            return false;
+        }
+        if let Some(mut round) = Round::new(proof.v, proof.t, setup.n_p, setup.sec_param) {
```

Comment:
This function has a lot of indentation which makes reading it hard. Let's instead short-circuit just like the above three lines.
```suggestion
        let mut round = match Round::new(proof.v, proof.t, setup.n_p, setup.sec_param) {
          Some(x) => x,
          None => return false,
        };
```
The same can be done in a couple more places below.